### ISSUE TITLE
Computation times are very different between two different versions of pymc3

### ISSUE THREAD

**Issue by tomicapretto** [pymc-devs/pymc#4009]
## Description of your problem

I want to fit a simple logistic regression in PyMC3 and I have tried with two different versions, 3.8 and 3.9.2. 
When I use the version 3.8, the sampling takes around 20 minutes, but when I use the latest version the sampling is below 10% of the objective after 20 minutes and the expected waiting time is around 4 hours.

I am using anaconda and I have uninstalled and installed everything from 0, but nothing changed. In both scenarios I used a different environment that only contains the packages required by the example.

```python

import numpy as np
import pandas as pd
import pymc3 as pm
from pandas.api.types import CategoricalDtype

# The original data handling process
data = pd.read_table(
    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
    header=None, sep = ",", skipinitialspace=True,
    names=['age', 'workclass', 'fnlwgt', 'education', 'education_num',
                  'marital_status', 'occupation', 'relationship', 'race', 'gender',
                  'captial_gain', 'capital_loss', 'hs_week', 'native_country','income']
)

categorical_cols = data.columns[data.dtypes == object].tolist()
for col in categorical_cols:
    data[col] = data[col].astype('category')
    
data = data[["age", "gender", "race", "education", "hs_week", "income"]]

education_mapping = {
    "1st-4th": "Primary", 
    "5th-6th": "Primary", 
    "7th-8th": "Primary", 
    "9th": "HS", 
    "10th": "HS", 
    "11th": "HS", 
    "12th": "HS",
    "HS-grad": "HS grad", 
    "Some-college": "Some college", 
    "Assoc-acdm": "Associate", 
    "Assoc-voc": "Associate", 
    "Bachelors": "Bachelors", 
    "Masters": "Masters", 
    "Doctorate": "Doctorate",
    "Professional school": "Prof. school" 
}

data["education_grouped"] = data["education"].map(education_mapping)

education_category_type = CategoricalDtype(
    categories = ["Primary", "HS", "HS grad", "Some college",
                  "Associate", "Bachelors", "Masters", "Prof. school", "Doctorate"], 
    ordered=True)

data["education_grouped"] = data["education_grouped"].astype(education_category_type)

data["gender"] = np.where(data["gender"] == "Female", 1, 0)
data["race"] = np.where(data["race"] == "White", 1, 0)
data["age"] = data["age"] / 10
data["age2"] = data["age"] ** 2
data["hs_week"] = data["hs_week"] / 10
data["hs_week2"] = data["hs_week"] ** 2
data["income"] = np.where(data["income"] == ">50K", 1, 0)

hs = ["HS grad", "Some college"]
undergrad = ["Associate", "Bachelors"]
grad = ["Masters", "Doctorate", "Prof. school"]

data["education_hs"] = np.where(data["education_grouped"].isin(hs), 1, 0)
data["education_undergrad"] = np.where(data["education_grouped"].isin(undergrad), 1, 0)
data["education_grad"] = np.where(data["education_grouped"].isin(grad), 1, 0)

# And the model
model_str = " + ".join((
    "income ~ age",
    "age2",
    "hs_week",
    "hs_week2",
    "gender",
    "race"
))

with pm.Model() as model:
    pm.glm.GLM.from_formula(
        formula = model_str,
        data = data,
        family = pm.glm.families.Binomial())
    trace = pm.sample(1000, tune=1000, init="adapt_diag")

# Finally, the session information.
from sinfo import sinfo
sinfo()
```
Output when PyMC3 version is 3.8   

```python
-----
numpy       1.18.5
pandas      1.0.5
pymc3       3.8
sinfo       0.3.1
-----
IPython             7.16.1
jupyter_client      6.1.5
jupyter_core        4.6.3
notebook            6.0.3
-----
Python 3.8.3 (default, Jul  2 2020, 17:30:36) [MSC v.1916 64 bit (AMD64)]
Windows-10-10.0.18362-SP0
4 logical CPU cores, AMD64 Family 23 Model 24 Stepping 1, AuthenticAMD
-----
Session information updated at 2020-07-11 17:55
```
![1](https://user-images.githubusercontent.com/25507629/87234169-c8d9ef00-c3a4-11ea-9e17-52250bc59ced.JPG)

Output when PyMC3 version is 3.9.2

```python
-----
numpy       1.18.5
pandas      1.0.5
pymc3       3.9.2
sinfo       0.3.1
-----
IPython             7.16.1
jupyter_client      6.1.5
jupyter_core        4.6.3
notebook            6.0.3
-----
Python 3.8.3 (default, Jul  2 2020, 17:30:36) [MSC v.1916 64 bit (AMD64)]
Windows-10-10.0.18362-SP0
4 logical CPU cores, AMD64 Family 23 Model 24 Stepping 1, AuthenticAMD
-----
Session information updated at 2020-07-11 18:25
```
I just terminated the execution before finishing before it was clear it was taking much longer.

![2](https://user-images.githubusercontent.com/25507629/87234170-cbd4df80-c3a4-11ea-9982-0db6bcadd6ff.JPG)

## Versions and main components

* PyMC3 Version: 3.8 and 3.9.2
* Theano Version: 1.0.4
* Python Version: 3.8.3
* Operating system: Windows 10
* How did you install PyMC3: pip

**Comment by nkaimcaudle** [comment]
Your code took about 5mins to complete on my machine at home, running debian, python 3.7.4, PyMC3 from master.

But at work I use Windows and I also noticed a large slowdown. I noticed this last Friday so don't have much more information at this stage.
Main environment: Win7, python 3.6, PyMC3 3.8, a script took ~30mins
New environment: Win7, python 3.7, PyMC3 3.9, same script took over 6hrs

**Comment by michaelosthege** [comment]
Is it just with NUTS, or are other samplers also impacted?

**Comment by nkaimcaudle** [comment]
At work my model uses just NUTS.

**Comment by aseyboldt** [comment]
Could you try to set `theano.config.mode = "FAST_RUN"` after import pymc3 but before sampling?

**Comment by tomicapretto** [comment]
> Is it just with NUTS, or are other samplers also impacted?

I've just tried with pm.Metropolis and I see the same difference. It took 3 mins for version 3.8 and around 30 mins for 3.9.2.

**Comment by tomicapretto** [comment]
> Could you try to set `theano.config.mode = "FAST_RUN"` after import pymc3 but before sampling?

I've just tried and nothing changed :/

**Comment by michaelosthege** [comment]
Here's the [diff for v3.8...v3.9.2](https://github.com/pymc-devs/pymc3/compare/v3.8...v3.9.2) let's skim through, maybe something fishy sticks out.

**Comment by aseyboldt** [comment]
Could you run the following code on both versions, so that we can compare the output? This prepares the function that computes the logp and the gradient, but also outputs some profiling info:

```
func = model.logp_dlogp_function(profile=True)
func.set_extra_values({})
np.random.seed(42)
x0 = np.random.randn(func.size)
for _ in range(1000):  # choose some number so that this takes a few seconds
    func(x0)
func.profile.summary()
```

**Comment by michaelosthege** [comment]
I'm running a short benchmark on different commits until I identified the one. This is what I have so far:


version | t_sampling
-- | --
3.8 | 38.03
af1b55f | 37.53
6f91999 | 40.83
803dbbef | 47.13
edca817 | 38.27
0ab48c6 | 37.56
4cb59429 | 266.47
d4d9ee3 | 260.72
fd72bd4 | 264.53
839206b | 258.69
3.9.1 | 263.01
3.9.2 | 257.0

I willl run your profiling script as soon as I found the digression commit.

**Comment by michaelosthege** [comment]
It's because of compiler flags: https://github.com/pymc-devs/pymc3/compare/0ab48c6...4cb59429

```
(pm3-dev2) C:\Users\osthege\Repos\pymc3-dev\benchmarks\benchmarks>python issue4009.py
Function profiling
==================
  Message: c:\users\osthege\repos\pymc3-dev\pymc3\model.py:555
  Time in 1000 calls to Function.__call__: 7.404492e+01s
  Time in Function.fn.__call__: 7.401392e+01s (99.958%)
  Time in thunks: 7.383091e+01s (99.711%)
  Total compile time: 2.320004e-01s
    Number of Apply nodes: 186
    Theano Optimizer time: 1.450155e-01s
       Theano validate time: 0.000000e+00s
    Theano Linker time (includes C, CUDA code generation/compiling): 4.600096e-02s
       Import time 0.000000e+00s
       Node make_thunk time 4.300380e-02s
           Node Elemwise{ge,no_inplace}(sigmoid.0, TensorConstant{(1,) of 0}) time 3.597045e-02s
           Node InplaceDimShuffle{}(Subtensor{int64}.0) time 1.031399e-03s
           Node Elemwise{sub,no_inplace}(hs_week, TensorConstant{0.0}) time 1.031160e-03s
           Node InplaceDimShuffle{}(Subtensor{int64}.0) time 1.003742e-03s
           Node InplaceDimShuffle{}((d__logp/d__logp_hs_week)) time 1.003027e-03s

Time in all call to theano.grad() 1.399994e-01s
Time since theano import 82.570s
Class
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  98.7%    98.7%      72.901s       6.45e-04s     Py  113000     113   theano.tensor.elemwise.Elemwise
   0.4%    99.2%       0.326s       1.63e-04s     Py    2000       2   theano.tensor.blas.Dot22
   0.2%    99.4%       0.167s       7.95e-06s     Py   21000      21   theano.tensor.subtensor.Subtensor
   0.2%    99.6%       0.130s       1.86e-05s     Py    7000       7   theano.tensor.subtensor.IncSubtensor
   0.1%    99.7%       0.093s       5.17e-06s     Py   18000      18   theano.tensor.elemwise.DimShuffle
   0.1%    99.8%       0.072s       5.14e-06s     Py   14000      14   theano.tensor.basic.Reshape
   0.1%    99.9%       0.062s       3.10e-05s     Py    2000       2   theano.tensor.elemwise.Sum
   0.1%   100.0%       0.060s       8.57e-06s     Py    7000       7   theano.compile.ops.Shape
   0.0%   100.0%       0.020s       1.00e-05s     Py    2000       2   theano.tensor.opt.MakeVector
   ... (remaining 0 Classes account for   0.00%(0.00s) of the runtime)

Ops
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  76.1%    76.1%      56.170s       5.62e-02s     Py    1000        1   sigmoid
  18.4%    94.4%      13.549s       2.71e-03s     Py    5000        5   Elemwise{second}
   1.1%    95.5%       0.800s       2.35e-05s     Py    34000       34   Elemwise{mul,no_inplace}
   0.8%    96.3%       0.597s       2.30e-05s     Py    26000       26   Elemwise{add,no_inplace}
   0.7%    97.0%       0.487s       2.43e-04s     Py    2000        2   Elemwise{log,no_inplace}
   0.6%    97.6%       0.467s       7.78e-05s     Py    6000        6   Elemwise{switch,no_inplace}
   0.4%    98.1%       0.326s       1.63e-04s     Py    2000        2   Dot22
   0.3%    98.4%       0.258s       1.84e-05s     Py    14000       14   Elemwise{true_div,no_inplace}
   0.3%    98.7%       0.212s       1.77e-05s     Py    12000       12   Elemwise{pow,no_inplace}
   0.3%    99.0%       0.209s       2.61e-05s     Py    8000        8   Elemwise{sub,no_inplace}
   0.2%    99.2%       0.130s       1.86e-05s     Py    7000        7   IncSubtensor{Inc;int64:int64:}
   0.2%    99.3%       0.117s       8.36e-06s     Py    14000       14   Subtensor{int64}
   0.1%    99.4%       0.062s       3.10e-05s     Py    2000        2   Sum{acc_dtype=float64}
   0.1%    99.5%       0.060s       8.57e-06s     Py    7000        7   Shape
   0.1%    99.6%       0.060s       3.00e-05s     Py    2000        2   Elemwise{eq,no_inplace}
   0.1%    99.6%       0.053s       4.08e-06s     Py    13000       13   InplaceDimShuffle{}
   0.1%    99.7%       0.050s       7.14e-06s     Py    7000        7   Subtensor{int64:int64:}
   0.1%    99.8%       0.045s       6.43e-06s     Py    7000        7   Reshape{1}
   0.0%    99.8%       0.034s       3.40e-05s     Py    1000        1   Elemwise{ge,no_inplace}
   0.0%    99.8%       0.030s       3.00e-05s     Py    1000        1   Elemwise{neg}
   ... (remaining 6 Ops account for   0.16%(0.11s) of the runtime)

Apply
------
<% time> <sum %> <apply time> <time per call> <#call> <id> <Apply name>
  76.1%    76.1%      56.170s       5.62e-02s   1000    19   sigmoid(Elemwise{add,no_inplace}.0)
   6.2%    82.3%       4.560s       4.56e-03s   1000    77   Elemwise{second}(Elemwise{switch,no_inplace}.0, InplaceDimShuffle{x}.0)
   6.1%    88.3%       4.470s       4.47e-03s   1000    82   Elemwise{second,no_inplace}(Elemwise{sub,no_inplace}.0, TensorConstant{(1,) of 0.0})
   6.0%    94.3%       4.445s       4.44e-03s   1000    88   Elemwise{second,no_inplace}(sigmoid.0, TensorConstant{(1,) of 0.0})
   0.4%    94.7%       0.280s       2.80e-04s   1000    21   Elemwise{log,no_inplace}(Elemwise{sub,no_inplace}.0)
   0.3%    95.0%       0.218s       2.18e-04s   1000    96   Dot22(TensorConstant{[[ 1.    1..    1.  ]]}, InplaceDimShuffle{0,x}.0)
   0.3%    95.3%       0.207s       2.07e-04s   1000    25   Elemwise{log,no_inplace}(sigmoid.0)
   0.1%    95.4%       0.108s       1.08e-04s   1000    16   Dot22(TensorConstant{[[ 1.    3..    1.  ]]}, InplaceDimShuffle{0,x}.0)
   0.1%    95.6%       0.102s       1.02e-04s   1000    78   Elemwise{Switch}(Elemwise{mul,no_inplace}.0, Elemwise{second}.0, TensorConstant{(1,) of 0.0})
   0.1%    95.7%       0.084s       8.40e-05s   1000    24   Elemwise{switch,no_inplace}(Elemwise{eq,no_inplace}.0, TensorConstant{[-inf -inf..-inf   0.]}, Elemwise{mul,no_inplace}.0)
   0.1%    95.8%       0.084s       8.40e-05s   1000    20   Elemwise{sub,no_inplace}(TensorConstant{(1,) of 1}, sigmoid.0)
   0.1%    95.9%       0.079s       7.90e-05s   1000    22   Elemwise{mul,no_inplace}(TensorConstant{[1 1 1 ... 1 1 0]}, Elemwise{log,no_inplace}.0)
   0.1%    96.0%       0.077s       7.70e-05s   1000    79   Elemwise{Switch}(Elemwise{eq,no_inplace}.0, TensorConstant{(1,) of 0.0}, Elemwise{Switch}.0)
   0.1%    96.1%       0.076s       7.60e-05s   1000    83   Elemwise{add,no_inplace}(Elemwise{second,no_inplace}.0, Elemwise{true_div}.0)
   0.1%    96.2%       0.072s       7.20e-05s   1000    85   Elemwise{Switch}(Elemwise{eq,no_inplace}.0, TensorConstant{(1,) of 0.0}, Elemwise{Switch}.0)
   0.1%    96.3%       0.070s       7.00e-05s   1000    26   Elemwise{mul,no_inplace}(TensorConstant{[0 0 0 ... 0 0 1]}, Elemwise{log,no_inplace}.0)
   0.1%    96.4%       0.069s       6.90e-05s   1000    37   Elemwise{switch,no_inplace}(Elemwise{mul,no_inplace}.0, Elemwise{add,no_inplace}.0, TensorConstant{(1,) of -inf})
   0.1%    96.5%       0.066s       6.60e-05s   1000    80   Elemwise{mul}(Elemwise{Switch}.0, TensorConstant{[1 1 1 ... 1 1 0]})
   0.1%    96.6%       0.063s       6.30e-05s   1000    86   Elemwise{mul}(Elemwise{Switch}.0, TensorConstant{[0 0 0 ... 0 0 1]})
   0.1%    96.7%       0.063s       6.30e-05s   1000    28   Elemwise{switch,no_inplace}(Elemwise{eq,no_inplace}.0, TensorConstant{[  0.   0...  0. -inf]}, Elemwise{mul,no_inplace}.0)
   ... (remaining 166 Apply instances account for 3.34%(2.47s) of the runtime)

Here are tips to potentially make your code run faster
                 (if you think of new ones, suggest them on the mailing list).
                 Test them first, as they are not guaranteed to always provide a speedup.
  - Try the Theano flag floatX=float32
  - Try installing amdlibm and set the Theano flag lib.amdlibm=True. This speeds up only some Elemwise operation.
-----
numpy       1.18.1
pandas      1.0.3
pymc3       3.8
sinfo       0.3.1
-----
Python 3.6.10 |Anaconda, Inc.| (default, May  7 2020, 19:46:08) [MSC v.1916 64 bit (AMD64)]
Windows-10-10.0.19041-SP0
6 logical CPU cores, Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
-----
Session information updated at 2020-07-13 13:21
```

**Comment by aseyboldt** [comment]
Ok, so it is the FAST_COMPILE then. We really need to get rid of that workaround...
It is interesting, that switching it back after import does not seem to fix the issue thought.

**Comment by tomicapretto** [comment]
As a final comment on my side, I just switched to Ubuntu and now it runs in ~8 minutes so I guess I'll keep using Ubuntu when doing this kind of stuff. Thank you all for the thorough responses :)

**Comment by nkaimcaudle** [comment]
If I manually comment out the line in __init__.py to change theano.config.mode then the script once agains runs in about 30mins. Thanks for the deep dive @michaelosthege

**Comment by twiecki** [comment]
Can't we just remove that line and be done with it?

**Comment by michaelosthege** [comment]
> Can't we just remove that line and be done with it?

I'll try tomorrow & run sampling tests and the Rtlive model to see if it helps..

**Comment by michaelosthege** [comment]
Looks like Adrian fixed it in c7afc215 already
